Disallow creation of tensors with duplicate names

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21795 Implement tensor.select(Dimname,int)
* #21789 Add at::dimname_to_position helper.
* #21752 named inference rule for tensor.select
* **#21781 Disallow creation of tensors with duplicate names**
* #21803 Rename Dimname::name to Dimname::full_name
* #21735 Copy NamedTensorMeta in TensorImpl::copy_tensor_data()
* #21648 Expose torch.empty(sizes, *, names, ...) to Python

Duplicate names must have tags and their tags must be different.
i.e., ("C.in", "C") is not allowed, but ("C.in", "C.out") is.

Test Plan:
- `python test/test_namedtensor.py -v` [namedtensor ci]

gh-metadata: pytorch pytorch 21781 gh/zou3519/51/head

Differential Revision: [D15833454](https://our.internmc.facebook.com/intern/diff/D15833454)

zou3519 committed:
Disallow creation of tensors with duplicate names

Duplicate names must have tags and their tags must be different.
i.e., ("C.in", "C") is not allowed, but ("C.in", "C.out") is.

zou3519 committed:
Update on "Disallow creation of tensors with duplicate names"

Disallow creation of tensors with duplicate names

Duplicate names must have tags and their tags must be different.
i.e., ("C.in", "C") is not allowed, but ("C.in", "C.out") is.

Test Plan:
- `python test/test_namedtensor.py -v` [namedtensor ci]

gh-metadata: pytorch pytorch 21781 gh/zou3519/51/head

nairbv commented on aten/src/ATen/NamedTensor.cpp:
```diff
@@ -12,6 +12,37 @@ bool NamedTensorMeta::has_names() const {
+      TORCH_CHECK(it->type() == NameType::TAGGED && dup->type() == NameType::TAGGED,
+          "Cannot construct a tensor with duplicate names unless they are tagged ",
+          "and have different tags. Got names: ", names, ", offending names: (",
+          *it, " and ", *dup, ").");
```
Will this catch `names=(C.in, C.in)`? it seems like it only checks that both are tagged, not that the tags are unique.

zou3519 replied:
The condition above will catch `names=(C.in, C.in)` and complain that the user is constructing a tensor with duplicate names

nairbv replied:
ah I missed that that's the full name.

Would it be better if Dimname attributes were `name` and `tag`, instead of `name` and `untagged_name`? Is  the full name is used more frequently?
Will there be cases where we need to extract the tag in isolation, meaning needing to parse the string from the name symbol?

zou3519 replied:
I don't see a case where we need to use the tag in isolation. The full name is going to be used pretty frequently (in `unify(Dimname, Dimname`).

I agree that "name", "tag" and "full name" get confusing to reason about. I don't know what a better naming scheme would be, though.

zou3519 replied:
Do you think it would be clearer if the Dimname attributes were `name` and `fullname` instead of `untagged_name` and `name`?

zou3519 replied:
I'm going to rename Dimname::name to Dimname::full_name for readability but keep `Dimname::untagged_name`

nairbv commented on aten/src/ATen/NamedTensor.cpp:
```diff
@@ -12,6 +12,37 @@ bool NamedTensorMeta::has_names() const {
+
+static void check_unique_names(DimnameList names) {
+  // Strategy: Compare each element with the ones that come after it.
+  // Although this is O(N^2), in practice N is small (no more than 25).
```
I don't love the n^2 here, and seems like it might be possible to write more efficient code that's also more concise, but I guess we can always optimize later.
If dimname defined an equality function that captured this specific kind of tagged uniqueness, we could presumably use std functions or set operators to check uniqueness. Maybe we should add such a functions to the dimname class?

zou3519 replied:
std::set constructs a red-black tree that is pretty heavy. I haven't benchmarked it but my hypothesis is that it'll be a lot slower.

Although it is O(n^2), it is effectively constant because the hypothesis is that users won't use unlimited numbers of dimensions.

zou3519 replied:
I feel like adding an equality function to Dimname is bad because we have a few types of equality and it's unclear which one the equality operator should take:
1) "name" equality: `name.name() == other.name()` && `name.untagged_name() == other.untagged_name()`
2) "match" equality: `name.untagged_name() == other.untagged_name()`
3) "untagged name uniqueness": what we're checking for here

I'm leaning toward an equality operator doing (1), but I think the semantics in code are clearer if there isn't an equality operator.

nairbv replied:
I meant something like `std::unordered_set` or `std::unique`, doesn't have to be the tree based version. If it weren't for the tagging logic we could just have something like this:
```
unordered_set<Symbol> s;
for_each(names.begin(), names.end(), [&](Dimname &n){ s.insert(n.untagged_name());})
s.size() == names.size();
```
Though there's some tradeoff I guess for allocating more memory to create a set, and I'm not sure if it can be done so concisely with the various tagging conditions.

I agree we wouldn't want to confuse the meaning of operator==() on Dimname, but we could define named equals methods (which I assumed can be passed as template args to set/unique operations in std). The 'equals' functions could each be given sensible names like `compatibleWith(Dimname other)` or `matches(Dimname other)` or whatever we think makes the most sense for each of the cases you describe.

zou3519 replied:
`std::unordered_set` is also not the most efficient thing in the world for small lists because of the hash function computation.

The thing about `std::unique` is that it modifies the vector in-place, so we'd need to copy it. In addition, it doesn't say which elements are unique, so we can't do the error checking that is currently implemented here (although there is an argument to be made that we don't need to tell the users what the offending elements were).

nairbv replied:
I'm not super familiar with C++ libraries, but I'm sure c++ has some efficient way to check for duplicate entries in a vector given a hash/equals method. I think the efficient hash operation on a `Symbol` would just be the address of the char*, which shouldn't really slow things down.

I imagine there are cases where someone may want to create large numbers of  named tensors (possibly all with the same names), in which case small efficiencies could be significant. I'm not sure about the tradeoff of memory allocation vs time complexity of small name lists though. We can always improve later.

Also I'm not super familiar with whats idiomatic in C++, but where we do have a few notions of equality for this particular type, even if we don't necessarily need a separate function for this particular case, I'd think it would be helpful to have those concepts of equality be be defined explicitly as members functions of the dimname class?

zou3519 replied:
> I'm not super familiar with C++ libraries, but I'm sure c++ has some efficient way to check for duplicate entries in a vector given a hash/equals method.

Not that I know of, but it's fine to do a std::unique first and then have the error reporting be slower (the user ran into an error, after all, and it won't be in the hot path). 

> Also I'm not super familiar with whats idiomatic in C++, but where we do have a few notions of equality for this particular type, even if we don't necessarily need a separate function for this particular case, I'd think it would be helpful to have those concepts of equality be be defined explicitly as members functions of the dimname class?

Yup, can do. That'll make it cleaner, thanks for pointing that out.

zou3519 committed:
Update on "Disallow creation of tensors with duplicate names"

Disallow creation of tensors with duplicate names

Duplicate names must have tags and their tags must be different.
i.e., ("C.in", "C") is not allowed, but ("C.in", "C.out") is.

Test Plan:
- `python test/test_namedtensor.py -v` [namedtensor ci]

gh-metadata: pytorch pytorch 21781 gh/zou3519/51/head

zou3519 committed:
Update on "Disallow creation of tensors with duplicate names"

Disallow creation of tensors with duplicate names

Duplicate names must have tags and their tags must be different.
i.e., ("C.in", "C") is not allowed, but ("C.in", "C.out") is.

Test Plan:
- `python test/test_namedtensor.py -v` [namedtensor ci]

gh-metadata: pytorch pytorch 21781 gh/zou3519/51/head
